torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3843,6 +3844,25 @@ def _make_data_dependent_error(self, expr, unhinted_expr, *, size_oblivious_resu
             # problem
         )
 
+    def _update_var_to_range(self, symbol, vr):
+        lower, upper = vr.lower, vr.upper
+
+        # If we have a size-like unbacked SymInt, refuse to refine the range to be
+        # less than two.  This is because when we intersect this range
+        # with [2, inf] for size oblivious tests, the range would be
+        # unsatisfiable.  In other words, once you have a size-like
+        # unbacked SymInt, we can never learn that it is exactly zero or one,
+        # because we would now give inconsistent results for all size
+        # oblivous tests!
+        if upper < 2 and symbol in self.size_like:
+            upper = 2
```

Comment:
upper -> lower?

Reply:
nope, we're trying to make sure the upper bound isn't too low here

Reply:
But this is bad. If `upper < 2`, that means we *know* that the variable cannot take the value 2, but here we are decreeing that it can!
We should do one of two things:
- Raise an error
- Remove the symbol from size_like.
I'm pretty sure the second option is not sound, so I think we should just raise.

None of them is a great solution, but that's what you get when you make some assumptions (sizelike-ness) and they do not turn out to be true!

Reply:
@lezcano It is sound, it is simply not complete, no? When we say that an unbacked symbol has some value range [min, max], we are not saying there *must* be a possible value that hits every element of the range (this would be incoherent). We are always allowed to relax the value range if necessary.

Reply:
When we have a `[min, max]` range it means that we know that it may take values in that range (perhaps not all of them), but *not outside that range*. As such, if `upper < 2` and it's size-like, we know that the size-like assumption was wrong.

Reply:
I don't understand the point you're trying to make. If we just think intuitively what it means for something to be size-like, the true range is from [0, inf]. If we discover that upper < 2, well, that's fine. There would only be a problem if upper < 0.

If in response to that, you say, "Well Ed, that's what the word size-like means, but in code, size like means [2, inf]", well, this is... not true? Like, when I have a size-like unbacked SymInt and at runtime I get x == 0 or x == 1, that's not an error. And then we get into the thing I was telling you were size oblivious + size-like unbacked gives you this alternative PyTorch frontend where you never look at the 0/1-ness of sizes...

Reply:
What you said is just partly true, I think. Your point, if I'm not mistaken is that: "since a value range is an upper bound on the range of values that a variable can take, setting `[0,2]` rather than `[0,1]` is a fair thing to do".

I really don't like this idea of "we have **proved** that a variable is 0 or 1 but we are completely discarding this information", not only that, but we now say that it is larger than 2" ?!

When we first discussed this feature, you said that it'd be rather rare to have an unbacked symint that we can prove that it's 0 or 1... but here we are.

In my opinion, the size oblivious semantics should be: "Given a size-like unbacked symint, we always assume that's >= 2 unless we can prove that it is 0 or 1". As discussed when we discussed this feature:
- It is true that the semantics of your program may change if our reasoning system gets better, but I argue that if it changes (i.e., we can prove that a variable is zero or one) then it is probably to enable optimisations. For example, you have a stride, and now we know that that stride is one and the tensor is actually contiguous.
- This gives the user all power. If you have a program where `x.contiguous()` is True in eager mode, and False in compiled mode, you can `assert` your way into making the two agree by giving enough preconditions to the system. With the current approach, you are stuck with the semantics being divergent with those form eager.

Reply:
> In my opinion, the size oblivious semantics should be: "Given a size-like unbacked symint, we always assume that's >= 2 unless we can prove that it is 0 or 1". As discussed when we discussed this feature:

But the specific problem here is that we have learned that the value range of the unbacked symint is `[0, 1]`. We don't know that it is zero or one specifically, so we can't refine it. You would only be able to strip size-like from it. But I do not want to do that, I think it is bad UX.

We might need to have a call about this.

Reply:
@lezcano I tried to condense my argument into this essay: https://docs.google.com/document/d/1ZSwf6x2DxR97eCvVcoNlYhzN_od7_kD_xwjr7gS1FtE/edit